_posts/2020-10-02-producer-tuning.md

```diff
@@ -0,0 +1,323 @@
+
+Compression is useful for improving throughput and reducing the load on storage, but might not be suitable for low latency applications where the cost of compression or decompression could be prohibitive. More on compression later.
+
+In our basic configuration, acknowledgments only confirm that messages have reached the broker, and there is no guarantee on message ordering.
+
+We can change that...
+
+### Key properties
+
+Let's look at how you can add or change producer configuration properties for fine tuning.
+
+Here are the properties we'll consider:
+
+* acks
+* min.insync.replicas (topics)
+* enable.idempotence
+* max.in.flight.requests.per.connection
+* retries
+* transactional.id
+* transaction.timeout.ms
+
+We won't discuss the properties in isolation here, as we'll be concentrating on how to use them to achieve a certain result.
+
+We'll look at how you can use a combination of these properties to regulate:
+
+* *Data durability* to minimize data loss when passing messages
+* *Ordered delivery* of messages
+* *Reliability guarantees* on message transactions involving batches of related messages
+* *Throughput* measured in the number of messages processed over a specific period
+* *Latency* measured in the time it takes for messages reach the broker
+
+It's quite likely you'll want to balance throughput and latency targets whilst also minimizing data loss and guaranteeing ordering.
+
+If you want to read more about what each property does, see Kafka's [Producer configs](https://kafka.apache.org/).
+
+Remember, the producer configuration properties you *can* use will also be driven by the requirements of your application. Avoid any change that breaks a property or guarantee provided by your application.
+
+### Data durability
+
+It you want to reduce the likelihood that messages are lost, use message delivery acknowledgments.
+
+Specify `acks=all` in your producer configuration to force a partition leader to replicate messages to a certain number of followers before acknowledging that the message request was successfully received.
+
+```properties
+# ...
+acks=all
+# ...
+```
+Use the `acks=all` producer configuration in conjunction with the `min.insync.replicas` property for topics. You set the `min.insync.replicas` property in the `KafkaTopic` resource.
+
+The `min.insync.replicas` configuration sets the numbers of brokers that need to have logged a message before an acknowledgment is sent to the producer.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1alpha1
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 1
+  replicas: 3
+  config:
+    min.insync.replicas: 2
+    #...
+```
+Using a topic replication factor of 3, and 2 in-sync replicas on other brokers, the producer can continue unaffected if a single broker is unavailable.
+
+If a second broker becomes unavailable, the producer won’t receive acknowledgments and won’t be able to produce more messages.
```

Comment:
It is true if the producer still sends with acks=all but it can still send if using the others type of acknowledgement.

Reply:
:+1: 

> If a second broker becomes unavailable, using `acks=all` the producer won’t receive acknowledgments and won’t be able to produce more messages.